feat(avm)!: Context id constraining in the tx trace

sirasistant opened the pull request:
Tracks context ids properly in the tx trace, and adds context id emission to tx_context events in simulation

dbanks12 commented on barretenberg/cpp/pil/vm2/tx.pil:
```diff
@@ -201,8 +201,16 @@ namespace tx;
 
-    pol commit context_id; // TODO: Constrain
-    pol commit next_context_id; // TODO: Constrain
+    // When sending a call to execution, next context id is incremented by one, since we use next_context_id for the call.
+    pol commit next_context_id_sent_to_enqueued_call;
+    // When coming back from a call, the next context id may have incremented an amount we can't predict from the tx trace.
+    pol commit next_context_id_from_enqueued_call;
+
+    // When issuing a call, context id is used for the call, making the next context id it plus one.
+    should_process_call_request * ((next_context_id + 1) - next_context_id_sent_to_enqueued_call) = 0;
+    // When comming back from the call, the newly available context id is the next_context_id that comes from the end of the call.
+    should_process_call_request * (next_context_id_from_enqueued_call - next_context_id') = 0;
+
```
Worth adding constraining tests for these?

sirasistant replied:
They have been included in the simple control flow test, which testes 2 setup, 1 app logic and 1 teardown call

sirasistant replied:
We still don't have a nice way to do constraining tests of the tx trace. I think I want to do some tx events builder for constraining tests or something like that hmmmm

dbanks12 reviewed (approved):
LGTM! One comment

sirasistant committed:
feat(avm)!: Context id constraining in the tx trace

Tracks context ids properly in the tx trace, and adds context id emission to tx_context events in simulation
